diff --git a/i18n_node/i18n_node.module b/i18n_node/i18n_node.module
index 8777aff..b6957e8 100644
--- a/i18n_node/i18n_node.module
+++ b/i18n_node/i18n_node.module
@@ -260,7 +260,7 @@ function i18n_node_permission() {
 function i18n_node_node_view($node, $view_mode, $langcode) {
   if (i18n_node_type_enabled($node)) {
     $extra_fields_display_settings = field_extra_fields_get_display('node', $node->type, $view_mode);
-    if (!empty($extra_fields_display_settings['language']['visible'])) {
+    if ($extra_fields_display_settings['language']['visible']) {
       $node->content['language'] = array(
         '#type' => 'item',
         '#title' => t('Language'),
